src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,260 @@
+ * - We can safely name the drivers as WiFiDriver / ThreadDriver, it should not be ambiguous for most cases
+ * - We can safely name the Status enum to Status, and some other structs -- if we are using using, then we should in the context of
+ *   writing something dedicated to network commissioning, then a single word Status should be clear enough.
+ */
+namespace NetworkCommissioning {
+
+constexpr size_t kMaxNetworkIDLen = 32;
+
+// TODO: This is exactly the same as the one in GroupDataProvider, this could be moved to src/lib/support
+template <typename T>
+class Iterator
+{
+public:
+    virtual ~Iterator() = default;
+    /**
+     *  @retval The number of entries in total that will be iterated.
+     */
+    virtual size_t Count() = 0;
+    /**
+     *   @param[out] item  Value associated with the next element in the iteration.
+     *  @retval true if the next entry is successfully retrieved.
+     *  @retval false if no more entries can be found.
+     */
+    virtual bool Next(T & item) = 0;
+    /**
+     * Release the memory allocated by this iterator.
+     * Must be called before the pointer goes out of scope.
+     */
+    virtual void Release() = 0;
+
+protected:
+    Iterator() = default;
+};
+
+/**
+ * The content should match the one in zap_generated/cluster-objects.h
+ */
+enum class Status : uint8_t
+{
+    kSuccess                = 0x00,
+    kOutOfRange             = 0x01,
+    kBoundsExceeded         = 0x02,
+    kNetworkIDNotFound      = 0x03,
+    kDuplicateNetworkID     = 0x04,
+    kNetworkNotFound        = 0x05,
+    kRegulatoryError        = 0x06,
+    kAuthFailure            = 0x07,
+    kUnsupportedSecurity    = 0x08,
+    kOtherConnectionFailure = 0x09,
+    kIPV6Failed             = 0x0A,
+    kIPBindFailed           = 0x0B,
+    kUnknownError           = 0x0C,
+};
+
+// The following structs follows the generated cluster object structs.
+struct Network
+{
+    uint8_t networkID[kMaxNetworkIDLen];
+    uint8_t networkIDLen;
+    bool connected;
+};
+
+struct WiFiScanResponse
+{
+public:
+    uint8_t security;
+    uint8_t ssid[DeviceLayer::Internal::kMaxWiFiSSIDLength];
+    uint8_t ssidLen;
+    uint8_t bssid[6];
+    uint8_t channel;
+    uint32_t wiFiBand;
+    int8_t rssi;
+};
+
+struct ThreadScanResponse
+{
+    uint64_t panId;
+    uint64_t extendedPanId;
+    char networkName[16];
+    uint8_t networkNameLen;
+    uint8_t ssidLen;
+    uint16_t channel;
+    uint8_t version;
+    uint64_t extendedAddress;
+    int8_t rssi;
+    uint8_t lqi;
+};
+
+using NetworkIterator            = Iterator<Network>;
+using WiFiScanResponseIterator   = Iterator<WiFiScanResponse>;
+using ThreadScanResponseIterator = Iterator<ThreadScanResponse>;
+
+// BaseDriver and WirelessDriver are the common interfaces for a network driver, platform drivers should not implement this
+// directly, instead, users are expected to implement WiFiDriver, ThreadDriver and EthernetDriver.
+namespace Internal {
+class BaseDriver
```

Comment:
Suggest adding Init() and a Shutdown() methods that are called at startup/shutdown from cluster init/finalize code.

This would allow cleanly loading intermediate state at init, such as setting the mRunningNetwork value to the correct data based on prior commissioning operations.